NPE on simulator launch

<!--
Verify first that your issue/request is not already reported on GitHub.
Also test if the latest release and main branch are affected too.
Always add information AFTER of these HTML comments, but no need to delete the comments.
-->

##### ISSUE TYPE
<!-- Pick one below and delete the rest -->
 * Bug Report

##### COMPONENT NAME
<!--
Categorize the issue, e.g. API, VR, VPN, UI, etc.
-->
~~~
API
~~~

##### CLOUDSTACK VERSION
<!--
New line separated list of affected versions, commit ID for issues on main branch.
-->

~~~
4.16.1.0-SNAPSHOT
~~~

##### CONFIGURATION
<!--
Information about the configuration if relevant, e.g. basic network, advanced networking, etc.  N/A otherwise
-->
I'm using the latest 4.16 Master and building the simulator image manually

##### OS / ENVIRONMENT
<!--
Information about the environment if relevant, N/A otherwise
-->
Ubuntu 21.10
OpenJDK 11

##### SUMMARY
<!-- Explain the problem/feature briefly -->
```bash
INFO  [o.a.c.s.d.p.DefaultHostListener] (ApiServer-10:ctx-d982031c ctx-6062c5c0) (logid:aac2a26e) Connection established between storage pool org.apache.cloud
stack.storage.datastore.PrimaryDataStoreImpl@3953e834 and host 1                                                                                              
INFO  [o.a.c.s.d.l.SimulatorImageStoreLifeCycleImpl] (ApiServer-1:ctx-58b428a9 ctx-9c2836b5) (logid:5b4df9e5) Trying to add a new data store at nfs://10.147.2
8.6:/export/home/sandbox/sstor to data center 1                                                                                                               
ERROR [o.a.c.s.i.TemplateServiceImpl] (ApiServer-1:ctx-58b428a9 ctx-9c2836b5) (logid:5b4df9e5) No remote endpoint to send command, check if host or ssvm is do
wn?                                                                                                                                                           
ERROR [c.c.u.SystemVmTemplateRegistration] (ApiServer-1:ctx-58b428a9 ctx-9c2836b5) (logid:5b4df9e5) Failed to register template for hypervisor: Simulator
java.lang.NullPointerException                                                                                                                                
        at com.cloud.upgrade.SystemVmTemplateRegistration.registerTemplate(SystemVmTemplateRegistration.java:676)            
        at com.cloud.storage.StorageManagerImpl$3.doInTransactionWithoutResult(StorageManagerImpl.java:2834)              
        at com.cloud.utils.db.TransactionCallbackNoReturn.doInTransaction(TransactionCallbackNoReturn.java:25)                                      
        at com.cloud.utils.db.Transaction$2.doInTransaction(Transaction.java:50)                                                                              
        at com.cloud.utils.db.Transaction.execute(Transaction.java:40)                                                                                        
        at com.cloud.utils.db.Transaction.execute(Transaction.java:47) 
        at com.cloud.storage.StorageManagerImpl.registerSystemVmTemplateOnFirstNfsStore(StorageManagerImpl.java:2791)                                         
        at com.cloud.storage.StorageManagerImpl.discoverImageStore(StorageManagerImpl.java:2775)                                                       
```

##### STEPS TO REPRODUCE
<!--

For new features, show how the feature would be used.
-->
I'm following the simulator docs [here](https://docs.cloudstack.apache.org/en/4.16.0.0/developersguide/developer_guide.html?highlight=simulator#using-the-simulator)
<!-- Paste example playbooks or commands between quotes below -->
~~~

~~~

<!-- You can also paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
<!-- What did you expect to happen when running the steps above? -->

~~~
Simulator starts successfully
~~~

##### ACTUAL RESULTS
<!-- What actually happened? -->

<!-- Paste verbatim command output between quotes below -->
~~~
Simulator fails with NPE on start and ACS won't boot
~~~

@tampler I built a new docker image from the 4.16 branch and observed the same behavior as reported before - I noticed the NPE, but MS is up and running
```
[17:30:41] pearl@pearl-XPS-15-7590:~/lab/cloudstack$ docker image ls
REPOSITORY                                                TAG       IMAGE ID       CREATED          SIZE
cloudstack/simulator                                      latest    d8640fffd454   31 minutes ago   13GB


[17:32:16] pearl@pearl-XPS-15-7590:~/lab/cloudstack$ docker ps
CONTAINER ID   IMAGE                         COMMAND                  CREATED          STATUS          PORTS                                                           NAMES
a8b95f3944f7   cloudstack/simulator:latest   "/usr/bin/supervisord"   10 minutes ago   Up 10 minutes   8080/tcp, 8096/tcp, 0.0.0.0:8080->5050/tcp, :::8080->5050/tcp   sim


### Grepping for the NPE & the subsequent logs inside the container
root@a8b95f3944f7:~# cat vmops.log | grep NullPointerException -A30
java.lang.NullPointerException
	at com.cloud.upgrade.SystemVmTemplateRegistration.registerTemplate(SystemVmTemplateRegistration.java:676)
	at com.cloud.storage.StorageManagerImpl$3.doInTransactionWithoutResult(StorageManagerImpl.java:2845)
	at com.cloud.utils.db.TransactionCallbackNoReturn.doInTransaction(TransactionCallbackNoReturn.java:25)
	at com.cloud.utils.db.Transaction$2.doInTransaction(Transaction.java:50)
	at com.cloud.utils.db.Transaction.execute(Transaction.java:40)
	at com.cloud.utils.db.Transaction.execute(Transaction.java:47)
	at com.cloud.storage.StorageManagerImpl.registerSystemVmTemplateOnFirstNfsStore(StorageManagerImpl.java:2802)
	at com.cloud.storage.StorageManagerImpl.discoverImageStore(StorageManagerImpl.java:2786)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:215)
	at com.sun.proxy.$Proxy114.discoverImageStore(Unknown Source)
	at org.apache.cloudstack.api.command.admin.storage.AddImageStoreCmd.execute(AddImageStoreCmd.java:135)
	at com.cloud.api.ApiDispatcher.dispatch(ApiDispatcher.java:156)
	at com.cloud.api.ApiServer.queueCommand(ApiServer.java:772)
	at com.cloud.api.ApiServer.handleRequest(ApiServer.java:596)
	at com.cloud.api.ApiServer.handle(ApiServer.java:502)
	at org.apache.http.protocol.HttpService.doService(HttpService.java:437)
	at org.apache.http.protocol.HttpService.handleRequest(HttpService.java:342)
	at com.cloud.api.ApiServer$WorkerTask.runInContext(ApiServer.java:1326)
	at org.apache.cloudstack.managed.context.ManagedContextRunnable$1.run(ManagedContextRunnable.java:48)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext$1.call(DefaultManagedContext.java:55)
	at org.apache.cloudstack.managed.context.impl.DefaultManagedContext.callWithContext(DefaultManagedContext.java:102)
--
Caused by: java.lang.NullPointerException
	at com.cloud.upgrade.SystemVmTemplateRegistration.registerTemplate(SystemVmTemplateRegistration.java:676)
	at com.cloud.storage.StorageManagerImpl$3.doInTransactionWithoutResult(StorageManagerImpl.java:2845)
	at com.cloud.utils.db.TransactionCallbackNoReturn.doInTransaction(TransactionCallbackNoReturn.java:25)
	... 32 more
2022-02-07 08:56:24,131 INFO  [c.c.u.SystemVmTemplateRegistration] (ApiServer-8:ctx-68a2e5b0 ctx-e497c9bd) (logid:b8cb6485) Unmounting store
2022-02-07 08:56:24,131 DEBUG [c.c.u.s.Script] (ApiServer-8:ctx-68a2e5b0 ctx-e497c9bd) (logid:b8cb6485) Executing: /bin/bash -c sudo umount /tmp/tmp10742134008486500295 
2022-02-07 08:56:24,132 DEBUG [c.c.u.s.Script] (ApiServer-8:ctx-68a2e5b0 ctx-e497c9bd) (logid:b8cb6485) Executing while with timeout : 3600000
2022-02-07 08:56:24,139 DEBUG [c.c.u.s.Script] (ApiServer-8:ctx-68a2e5b0 ctx-e497c9bd) (logid:b8cb6485) Exit value is 32
2022-02-07 08:56:24,139 DEBUG [c.c.u.s.Script] (ApiServer-8:ctx-68a2e5b0 ctx-e497c9bd) (logid:b8cb6485) umount: /tmp/tmp10742134008486500295: no mount point specified.
2022-02-07 08:56:24,172 INFO  [c.c.c.ConfigurationManagerImpl] (ApiServer-9:ctx-3056b89e ctx-70b4389a) (logid:8d178580) No storage traffic type was specified by admin, create default storage traffic on physical network 200 with same configure of management traffic type
2022-02-07 08:56:24,415 DEBUG [c.c.a.ApiServlet] (qtp1665366772-287:ctx-e669acd4) (logid:2f8314ef) ===START===  127.0.0.1 -- POST  
2022-02-07 08:56:24,430 DEBUG [c.c.u.AccountManagerImpl] (qtp1665366772-287:ctx-e669acd4) (logid:2f8314ef) Attempting to log in user: admin in domain 1
2022-02-07 08:56:24,432 DEBUG [o.a.c.s.a.PBKDF2UserAuthenticator] (qtp1665366772-287:ctx-e669acd4) (logid:2f8314ef) Retrieving user: admin
2022-02-07 08:56:24,708 DEBUG [c.c.u.AccountManagerImpl] (qtp1665366772-287:ctx-e669acd4) (logid:2f8314ef) CIDRs from which account 'Acct[149f6018-87c4-11ec-81c5-50eb7122da94-admin] -- Account {"id": 2, "name": "admin", "uuid": "149f6018-87c4-11ec-81c5-50eb7122da94"}' is allowed to perform API calls: 0.0.0.0/0,::/0

```
And I can access the UI:

![image](https://user-images.githubusercontent.com/10495417/152785042-f3bc50cd-a563-4bef-bb35-b4c598303af8.png)

--- a/server/src/main/java/com/cloud/storage/StorageManagerImpl.java
+++ b/server/src/main/java/com/cloud/storage/StorageManagerImpl.java
@@ -2808,32 +2808,33 @@ public class StorageManagerImpl extends ManagerBase implements StorageManager, C
                             Pair<String, Long> storeUrlAndId = new Pair<>(url, store.getId());
                             for (HypervisorType hypervisorType : hypSet) {
                                 try {
-                                    if (HypervisorType.Simulator != hypervisorType) {
-                                        String templateName = getValidTemplateName(zoneId, hypervisorType);
-                                        Pair<Hypervisor.HypervisorType, String> hypervisorAndTemplateName =
-                                                new Pair<>(hypervisorType, templateName);
-                                        Long templateId = systemVmTemplateRegistration.getRegisteredTemplateId(hypervisorAndTemplateName);
-                                        VMTemplateVO vmTemplateVO = null;
-                                        TemplateDataStoreVO templateVO = null;
-                                        if (templateId != null) {
-                                            vmTemplateVO = _templateDao.findById(templateId);
-                                            templateVO = _templateStoreDao.findByTemplate(templateId, DataStoreRole.Image);
-                                            if (templateVO != null) {
-                                                try {
-                                                    if (SystemVmTemplateRegistration.validateIfSeeded(url, templateVO.getInstallPath())) {
-                                                        continue;
-                                                    }
-                                                } catch (Exception e) {
-                                                    s_logger.error("Failed to validated if template is seeded", e);
+                                    if (HypervisorType.Simulator == hypervisorType) {
+                                        continue;
+                                    }
+                                    String templateName = getValidTemplateName(zoneId, hypervisorType);
+                                    Pair<Hypervisor.HypervisorType, String> hypervisorAndTemplateName =
+                                            new Pair<>(hypervisorType, templateName);
+                                    Long templateId = systemVmTemplateRegistration.getRegisteredTemplateId(hypervisorAndTemplateName);
+                                    VMTemplateVO vmTemplateVO = null;
+                                    TemplateDataStoreVO templateVO = null;
+                                    if (templateId != null) {
+                                        vmTemplateVO = _templateDao.findById(templateId);
+                                        templateVO = _templateStoreDao.findByTemplate(templateId, DataStoreRole.Image);
+                                        if (templateVO != null) {
+                                            try {
+                                                if (SystemVmTemplateRegistration.validateIfSeeded(url, templateVO.getInstallPath())) {
+                                                    continue;
                                                 }
+                                            } catch (Exception e) {
+                                                s_logger.error("Failed to validated if template is seeded", e);
                                             }
                                         }
-                                        SystemVmTemplateRegistration.mountStore(storeUrlAndId.first(), filePath);
-                                        if (templateVO != null && vmTemplateVO != null) {
-                                            systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, vmTemplateVO, filePath);
-                                        } else {
-                                            systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, filePath);
-                                        }
+                                    }
+                                    SystemVmTemplateRegistration.mountStore(storeUrlAndId.first(), filePath);
+                                    if (templateVO != null && vmTemplateVO != null) {
+                                        systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, vmTemplateVO, filePath);
+                                    } else {
+                                        systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, filePath);
                                     }
                                 } catch (CloudRuntimeException e) {
                                     SystemVmTemplateRegistration.unmountStore(filePath);

--- a/server/src/main/java/com/cloud/storage/StorageManagerImpl.java
+++ b/server/src/main/java/com/cloud/storage/StorageManagerImpl.java
@@ -2808,30 +2808,32 @@ public class StorageManagerImpl extends ManagerBase implements StorageManager, C
                             Pair<String, Long> storeUrlAndId = new Pair<>(url, store.getId());
                             for (HypervisorType hypervisorType : hypSet) {
                                 try {
-                                    String templateName = getValidTemplateName(zoneId, hypervisorType);
-                                    Pair<Hypervisor.HypervisorType, String> hypervisorAndTemplateName =
-                                            new Pair<>(hypervisorType, templateName);
-                                    Long templateId = systemVmTemplateRegistration.getRegisteredTemplateId(hypervisorAndTemplateName);
-                                    VMTemplateVO vmTemplateVO = null;
-                                    TemplateDataStoreVO templateVO = null;
-                                    if (templateId != null) {
-                                        vmTemplateVO = _templateDao.findById(templateId);
-                                        templateVO = _templateStoreDao.findByTemplate(templateId, DataStoreRole.Image);
-                                        if (templateVO != null) {
-                                            try {
-                                                if (SystemVmTemplateRegistration.validateIfSeeded(url, templateVO.getInstallPath())) {
-                                                    continue;
+                                    if (HypervisorType.Simulator != hypervisorType) {
+                                        String templateName = getValidTemplateName(zoneId, hypervisorType);
+                                        Pair<Hypervisor.HypervisorType, String> hypervisorAndTemplateName =
+                                                new Pair<>(hypervisorType, templateName);
+                                        Long templateId = systemVmTemplateRegistration.getRegisteredTemplateId(hypervisorAndTemplateName);
+                                        VMTemplateVO vmTemplateVO = null;
+                                        TemplateDataStoreVO templateVO = null;
+                                        if (templateId != null) {
+                                            vmTemplateVO = _templateDao.findById(templateId);
+                                            templateVO = _templateStoreDao.findByTemplate(templateId, DataStoreRole.Image);
+                                            if (templateVO != null) {
+                                                try {
+                                                    if (SystemVmTemplateRegistration.validateIfSeeded(url, templateVO.getInstallPath())) {
+                                                        continue;
+                                                    }
+                                                } catch (Exception e) {
+                                                    s_logger.error("Failed to validated if template is seeded", e);
                                                 }
-                                            } catch (Exception e) {
-                                                s_logger.error("Failed to validated if template is seeded", e);
                                             }
                                         }
-                                    }
-                                    SystemVmTemplateRegistration.mountStore(storeUrlAndId.first(), filePath);
-                                    if (templateVO != null && vmTemplateVO != null) {
-                                        systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, vmTemplateVO, filePath);
-                                    } else {
-                                        systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, filePath);
+                                        SystemVmTemplateRegistration.mountStore(storeUrlAndId.first(), filePath);
+                                        if (templateVO != null && vmTemplateVO != null) {
+                                            systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, vmTemplateVO, filePath);
+                                        } else {
+                                            systemVmTemplateRegistration.registerTemplate(hypervisorAndTemplateName, storeUrlAndId, filePath);
+                                        }
                                     }
                                 } catch (CloudRuntimeException e) {
                                     SystemVmTemplateRegistration.unmountStore(filePath);

